crates/assists/src/utils/insert_use.rs

```diff
@@ -22,525 +18,566 @@ pub(crate) fn find_insert_use_container(
 ) -> Option<Either<ast::ItemList, ast::SourceFile>> {
     ctx.sema.ancestors_with_macros(position.clone()).find_map(|n| {
         if let Some(module) = ast::Module::cast(n.clone()) {
-            return module.item_list().map(|it| Either::Left(it));
+            return module.item_list().map(Either::Left);
         }
         Some(Either::Right(ast::SourceFile::cast(n)?))
     })
 }
 
-/// Creates and inserts a use statement for the given path to import.
-/// The use statement is inserted in the scope most appropriate to the
-/// the cursor position given, additionally merged with the existing use imports.
 pub(crate) fn insert_use_statement(
     // Ideally the position of the cursor, used to
     position: &SyntaxNode,
     path_to_import: &str,
-    ctx: &AssistContext,
-    builder: &mut TextEditBuilder,
+    ctx: &crate::assist_context::AssistContext,
+    builder: &mut text_edit::TextEditBuilder,
 ) {
-    let target = path_to_import.split("::").map(SmolStr::new).collect::<Vec<_>>();
-    let container = find_insert_use_container(position, ctx);
+    insert_use(position.clone(), make::path_from_text(path_to_import), Some(MergeBehaviour::Full));
+}
+
+pub fn insert_use(
+    where_: SyntaxNode,
+    path: ast::Path,
+    merge_behaviour: Option<MergeBehaviour>,
+) -> SyntaxNode {
+    let use_item = make::use_(make::use_tree(path.clone(), None, None, false));
+    // merge into existing imports if possible
+    if let Some(mb) = merge_behaviour {
+        for existing_use in where_.children().filter_map(ast::Use::cast) {
+            if let Some(merged) = try_merge_imports(&existing_use, &use_item, mb) {
+                let to_delete: SyntaxElement = existing_use.syntax().clone().into();
+                let to_delete = to_delete.clone()..=to_delete;
+                let to_insert = iter::once(merged.syntax().clone().into());
+                return algo::replace_children(&where_, to_delete, to_insert);
+            }
+        }
+    }
+
+    // either we weren't allowed to merge or there is no import that fits the merge conditions
+    // so look for the place we have to insert to
+    let (insert_position, add_blank) = find_insert_position(&where_, path);
+
+    let to_insert: Vec<SyntaxElement> = {
+        let mut buf = Vec::new();
+
+        if add_blank == AddBlankLine::Before {
+            buf.push(make::tokens::single_newline().into());
+        }
+
+        buf.push(use_item.syntax().clone().into());
+
+        if add_blank == AddBlankLine::After {
+            buf.push(make::tokens::single_newline().into());
+        } else if add_blank == AddBlankLine::AfterTwice {
+            buf.push(make::tokens::single_newline().into());
+            buf.push(make::tokens::single_newline().into());
```

Comment:
`make::tokens::blank_line()` maybe?

Reply:
I'm honestly not sure how I would implement that function, I'm not too used to how `rowan` and the like works yet. So I'm unsure how I would fetch two newlines as a single `SyntaxToken`.